pkg/lib/sms/twilio/smsclient.go

```diff
@@ -147,17 +158,68 @@ func (t *TwilioClient) Send(ctx context.Context, options *smsclient.SendOptions)
 		sendCtx.Twilio.SegmentCount = segmentCount
 	})
 
-	// Success case.
-	if sendSMSResponse.ErrorCode == nil {
-		return &smsclient.SendResultSuccess{
-			DumpedResponse: dumpedResponse,
-		}, nil
+	return &smsclient.SendResultSuccess{
+		DumpedResponse: dumpedResponse,
+	}, nil
+
+}
+
+func (t *TwilioClient) parseAndHandleErrorResponse(
+	responseBody []byte,
+	dumpedResponse []byte,
+) error {
+	errResponse, err := ParseErrorResponse(responseBody)
+
+	if err != nil {
+		var jsonUnmarshalErr *json.UnmarshalTypeError
+		if errors.As(err, &jsonUnmarshalErr) {
+			// Not something we can understand, return an error with the dumped response
+			return &smsclient.SendResultError{
+				DumpedResponse: dumpedResponse,
+			}
+		} else {
+			return errors.Join(err, &smsclient.SendResultError{
+				DumpedResponse: dumpedResponse,
+			})
+		}
 	}
 
-	// Failed case.
-	return nil, &smsclient.SendResultError{
+	return t.makeError(errResponse.Code, dumpedResponse)
+}
+
+func (t *TwilioClient) makeError(
+	errorCode int,
+	dumpedResponse []byte,
+) *smsclient.SendResultError {
+	err := &smsclient.SendResultError{
 		DumpedResponse: dumpedResponse,
 	}
+
+	// See https://www.twilio.com/docs/api/errors
+	switch errorCode {
+	case 21211:
+		err.Code = api.CodeInvalidPhoneNumber
+		err.ErrorDetail = fmt.Sprintf("%d", errorCode)
```

Comment:
Now I see what is error_detail. But I doubt whether it is suitable for showing to end-users.

We already have `error_description` which is `error.Error()`. What about we introduce a TwilioError that is added the error chain. Then the original `error_description` will include the information in the TwilioError.

For the information in TwilioError, I think we should just include unparsed JSON body. Here is what I mean

```
type TwilioError struct {
  HTTPBody []byte `json:"http_body,omitempty"`
  ErrorCode *int `json:"error_code,omitempty"`
}

func (e *TwilioError) Error() string {
  // Use json.Marshal to marshal itself.
}
```

Reply:
My intention is to provide info for developers of the project to know about the actual error returned by the sms gateway easily in the ui, without having to contact us to read the logs. This cannot be done by `error_description` because it might include information we cannot disclose to end user.

Reply:
As discussed this is not needed